File: src/services/api/vesselService.ts
/**
 * Vessel Service - API calls for vessel management
 */

import apiClient from './apiClient';
import type { Vessel, VesselCreate, VesselDocument, VesselDID, VesselDocumentType } from '@/types';

interface VesselApiResponse {
  id: string;
  imo: string;
  name: string;
  type: string;
  flag: string;
  gross_tonnage: number;
  year_built: number;
  did: string | null;
  owner_did: string;
  status: string;
  documents: Array<{
    id: string;
    vessel_id: string;
    type: string;
    file_name: string;
    status: string;
    ipfs_cid: string | null;
    uploaded_at: string;
    verified_at: string | null;
  }>;
  ownership_credential_id: string | null;
  created_at: string;
  updated_at: string | null;
}

function transformVesselResponse(response: VesselApiResponse): Vessel {
  return {
    id: response.id,
    imo: response.imo,
    name: response.name,
    type: response.type as Vessel['type'],
    flag: response.flag,
    grossTonnage: response.gross_tonnage,
    yearBuilt: response.year_built,
    did: response.did || undefined,
    ownerDid: response.owner_did,
    status: response.status as Vessel['status'],
    documents: response.documents.map((doc) => ({
      id: doc.id,
      vesselId: doc.vessel_id,
      type: doc.type as VesselDocumentType,
      fileName: doc.file_name,
      status: doc.status as VesselDocument['status'],
      ipfsCid: doc.ipfs_cid || undefined,
      uploadedAt: doc.uploaded_at,
      verifiedAt: doc.verified_at || undefined,
    })),
    ownershipCredentialId: response.ownership_credential_id || undefined,
    createdAt: response.created_at,
    updatedAt: response.updated_at || undefined,
  };
}

export const vesselService = {
  /**
   * List all vessels, optionally filtered by owner DID
   */
  listVessels: async (ownerDid?: string): Promise<Vessel[]> => {
    const endpoint = ownerDid
      ? `/vessels/?owner_did=${encodeURIComponent(ownerDid)}`
      : '/vessels/';
    const response = await apiClient.get<VesselApiResponse[]>(endpoint);
    return response.map(transformVesselResponse);
  },

  /**
   * Create a new vessel
   */
  createVessel: async (data: VesselCreate): Promise<Vessel> => {
    const request = {
      imo: data.imo,
      name: data.name,
      type: data.type,
      flag: data.flag,
      gross_tonnage: data.grossTonnage,
      year_built: data.yearBuilt,
      owner_did: data.ownerDid,
    };
    const response = await apiClient.post<VesselApiResponse>('/vessels/', request);
    return transformVesselResponse(response);
  },

  /**
   * Get a vessel by ID
   */
  getVessel: async (vesselId: string): Promise<Vessel> => {
    const response = await apiClient.get<VesselApiResponse>(`/vessels/${vesselId}`);
    return transformVesselResponse(response);
  },

  /**
   * Get a vessel by IMO number
   */
  getVesselByIMO: async (imo: string): Promise<Vessel | null> => {
    try {
      const response = await apiClient.get<VesselApiResponse>(`/vessels/imo/${encodeURIComponent(imo)}`);
      return transformVesselResponse(response);
    } catch {
      return null;
    }
  },

  /**
   * Create a DID for a vessel (platform-controlled)
   */
  createVesselDID: async (vesselId: string): Promise<VesselDID> => {
    const response = await apiClient.post<{
      vessel_id: string;
      vessel_did: string;
      managed_by: string;
      created_at: string;
    }>(`/vessels/${vesselId}/did`);
    return {
      vesselId: response.vessel_id,
      vesselDid: response.vessel_did,
      managedBy: response.managed_by,
      createdAt: response.created_at,
    };
  },

  /**
   * Upload a document for a vessel
   */
  uploadDocument: async (
    vesselId: string,
    documentType: VesselDocumentType,
    fileName: string
  ): Promise<VesselDocument> => {
    const response = await apiClient.post<{
      id: string;
      vessel_id: string;
      type: string;
      file_name: string;
      status: string;
      ipfs_cid: string | null;
      uploaded_at: string;
      verified_at: string | null;
    }>(`/vessels/${vesselId}/documents`, {
      type: documentType,
      file_name: fileName,
    });
    return {
      id: response.id,
      vesselId: response.vessel_id,
      type: response.type as VesselDocumentType,
      fileName: response.file_name,
      status: response.status as VesselDocument['status'],
      ipfsCid: response.ipfs_cid || undefined,
      uploadedAt: response.uploaded_at,
      verifiedAt: response.verified_at || undefined,
    };
  },

  /**
   * Mark a document as verified
   */
  verifyDocument: async (vesselId: string, documentId: string): Promise<VesselDocument> => {
    const response = await apiClient.post<{
      id: string;
      vessel_id: string;
      type: string;
      file_name: string;
      status: string;
      ipfs_cid: string | null;
      uploaded_at: string;
      verified_at: string | null;
    }>(`/vessels/${vesselId}/documents/${documentId}/verify`);
    return {
      id: response.id,
      vesselId: response.vessel_id,
      type: response.type as VesselDocumentType,
      fileName: response.file_name,
      status: response.status as VesselDocument['status'],
      ipfsCid: response.ipfs_cid || undefined,
      uploadedAt: response.uploaded_at,
      verifiedAt: response.verified_at || undefined,
    };
  },

  /**
   * Update vessel status
   */
  updateVesselStatus: async (vesselId: string, status: Vessel['status']): Promise<Vessel> => {
    const response = await apiClient.put<VesselApiResponse>(
      `/vessels/${vesselId}/status?status=${status}`
    );
    return transformVesselResponse(response);
  },
};

export default vesselService;

File: backend/services/vessel_service.py
"""
Vessel service for managing vessel records and DIDs.
"""

from datetime import datetime
from typing import Optional, List
import uuid

from models.vessel import (
    Vessel, VesselCreate, VesselResponse, VesselDocument,
    VesselStatus, DocumentStatus, VesselDocumentType, DocumentUpload
)
from services.json_store import JSONStore
from config import VESSELS_FILE, PLATFORM_DID


class VesselService:
    """Service for vessel operations."""

    def __init__(self):
        self.store = JSONStore(VESSELS_FILE)

    def create_vessel(self, data: VesselCreate) -> VesselResponse:
        """Create a new vessel record."""
        # Check if vessel with same IMO already exists
        existing = self.store.get_by_field('imo', data.imo)
        if existing:
            return VesselResponse(**existing)

        vessel_id = f"VSL-{str(uuid.uuid4())[:8].upper()}"

        vessel = {
            'id': vessel_id,
            'imo': data.imo,
            'name': data.name,
            'type': data.type.value,
            'flag': data.flag,
            'gross_tonnage': data.gross_tonnage,
            'year_built': data.year_built,
            'did': None,  # Will be created by platform
            'owner_did': data.owner_did,
            'status': VesselStatus.PENDING_VERIFICATION.value,
            'documents': [],
            'ownership_credential_id': None,
            'created_at': datetime.utcnow().isoformat(),
            'updated_at': None
        }

        self.store.create(vessel)
        return VesselResponse(**vessel)

    def get_vessel(self, vessel_id: str) -> Optional[VesselResponse]:
        """Get a vessel by ID."""
        record = self.store.get_by_id(vessel_id)
        if record:
            return VesselResponse(**record)
        return None

    def get_vessel_by_imo(self, imo: str) -> Optional[VesselResponse]:
        """Get a vessel by IMO number."""
        record = self.store.get_by_field('imo', imo)
        if record:
            return VesselResponse(**record)
        return None

    def get_vessel_by_did(self, did: str) -> Optional[VesselResponse]:
        """Get a vessel by DID."""
        record = self.store.get_by_field('did', did)
        if record:
            return VesselResponse(**record)
        return None

    def list_vessels_by_owner(self, owner_did: str) -> List[VesselResponse]:
        """List all vessels owned by a specific shipowner."""
        records = self.store.find_by_field('owner_did', owner_did)
        return [VesselResponse(**r) for r in records]

    def list_all_vessels(self) -> List[VesselResponse]:
        """List all vessels."""
        records = self.store.get_all()
        return [VesselResponse(**r) for r in records]

    def create_vessel_did(self, vessel_id: str) -> Optional[dict]:
        """
        Create a DID for a vessel.
        Vessel DIDs are platform-controlled (neutral).
        Format: did:xrpl:1:rVESSEL{vessel_id}
        """
        vessel = self.store.get_by_id(vessel_id)
        if not vessel:
            return None

        if vessel.get('did'):
            # DID already exists
            return {
                'vessel_id': vessel_id,
                'vessel_did': vessel['did'],
                'managed_by': PLATFORM_DID,
                'created_at': vessel.get('created_at')
            }

        # Generate a vessel DID (platform-controlled)
        # In production, this would create an actual XRPL account
        vessel_did = f"did:xrpl:1:rVESSEL{vessel_id.replace('-', '')}"

        updated = self.store.update(vessel_id, {
            'did': vessel_did,
            'status': VesselStatus.VERIFIED.value
        })

        if updated:
            return {
                'vessel_id': vessel_id,
                'vessel_did': vessel_did,
                'managed_by': PLATFORM_DID,
                'created_at': datetime.utcnow().isoformat()
            }
        return None

    def add_document(self, vessel_id: str, doc: DocumentUpload) -> Optional[VesselDocument]:
        """Add a document to a vessel."""
        vessel = self.store.get_by_id(vessel_id)
        if not vessel:
            return None

        doc_id = f"DOC-{str(uuid.uuid4())[:8].upper()}"
        document = {
            'id': doc_id,
            'vessel_id': vessel_id,
            'type': doc.type.value,
            'file_name': doc.file_name,
            'status': DocumentStatus.UPLOADED.value,
            'ipfs_cid': None,
            'uploaded_at': datetime.utcnow().isoformat(),
            'verified_at': None
        }

        documents = vessel.get('documents', [])
        documents.append(document)

        # Update vessel status if documents submitted
        new_status = VesselStatus.DOCUMENTS_SUBMITTED.value if len(documents) >= 2 else vessel['status']

        self.store.update(vessel_id, {
            'documents': documents,
            'status': new_status
        })

        return VesselDocument(**document)

    def verify_document(self, vessel_id: str, document_id: str) -> Optional[VesselDocument]:
        """Mark a document as verified."""
        vessel = self.store.get_by_id(vessel_id)
        if not vessel:
            return None

        documents = vessel.get('documents', [])
        for i, doc in enumerate(documents):
            if doc['id'] == document_id:
                documents[i]['status'] = DocumentStatus.VERIFIED.value
                documents[i]['verified_at'] = datetime.utcnow().isoformat()
                self.store.update(vessel_id, {'documents': documents})
                return VesselDocument(**documents[i])
        return None

    def update_vessel_status(self, vessel_id: str, status: VesselStatus) -> Optional[VesselResponse]:
        """Update vessel status."""
        updated = self.store.update(vessel_id, {'status': status.value})
        if updated:
            return VesselResponse(**updated)
        return None

    def set_ownership_credential(self, vessel_id: str, credential_id: str) -> Optional[VesselResponse]:
        """Set the ownership credential ID for a vessel."""
        updated = self.store.update(vessel_id, {
            'ownership_credential_id': credential_id,
            'status': VesselStatus.ACTIVE.value
        })
        if updated:
            return VesselResponse(**updated)
        return None
